The Delta table schema is not the same as the Hive schema.

Why is it inconsistent? I checked that the schema of delta table and hive table are correct and the quantity is the same	


at org.apache.hadoop.util.RunJar.main(RunJar.java:227)
Caused by: MetaException(message:The Delta table schema is not the same as the Hive schema. Please update your Hive
table's schema to match the Delta table schema.
Delta table schema: StructType(StructField(types,StringType,true), StructField(tableName,StringType,true), StructField(primaryKey,StringType,true), StructField(timesmap,StringType,true), StructField(id,StringType,true), StructField(task_id,StringType,true), StructField(order_id,StringType,true), StructField(task_name,StringType,true), StructField(display_name,StringType,true), StructField(task_state,StringType,true), StructField(create_time,StringType,true), StructField(update_time,StringType,true), StructField(end_time,StringType,true), StructField(parent_task_id,StringType,true), StructField(variable,StringType,true), StructField(trade_id,StringType,true), StructField(partition,StringType,true))
Hive schema: struct<types:string,tablename:string,primarykey:string,timesmap:string,id:string,task_id:string,order_id:string,task_name:string,display_name:string,task_state:string,create_time:string,update_time:string,end_time:string,parent_task_id:string,variable:string,trade_id:string,partition:string>)
	at org.apache.spark.sql.delta.DeltaHelper$.metaInconsistencyException(DeltaHelper.scala:252)
	at org.apache.spark.sql.delta.DeltaHelper$.checkTableSchema(DeltaHelper.scala:176)
	at io.delta.hive.DeltaStorageHandler.preCreateTable(DeltaStorageHandler.scala:195)

--- a/hive/src/main/scala/org/apache/spark/sql/delta/DeltaHelper.scala
+++ b/hive/src/main/scala/org/apache/spark/sql/delta/DeltaHelper.scala
@@ -19,6 +19,8 @@ package org.apache.spark.sql.delta
 import java.net.URI
 
 import scala.collection.mutable
+import scala.collection.JavaConverters._
+
 import io.delta.hive.{DeltaStorageHandler, PartitionColumnInfo}
 import org.apache.hadoop.fs._
 import org.apache.hadoop.hive.metastore.api.MetaException
@@ -183,8 +185,8 @@ object DeltaHelper extends Logging {
         val sparkField = sparkStruct(i)
         val hiveFieldName = hiveStruct.getAllStructFieldNames.get(i)
         val hiveFieldType = hiveStruct.getAllStructFieldTypeInfos.get(i)
-        // TODO Do we need to respect case insensitive config?
-        sparkField.name == hiveFieldName && isSameType(sparkField.dataType, hiveFieldType)
+        sparkField.name.equalsIgnoreCase(hiveFieldName) &&
+          isSameType(sparkField.dataType, hiveFieldType)
       }
     } else {
       false
@@ -245,11 +247,20 @@ object DeltaHelper extends Logging {
   private def metaInconsistencyException(
       deltaSchema: StructType,
       hiveSchema: StructTypeInfo): MetaException = {
+    val hiveSchemaString = hiveSchema.getAllStructFieldNames
+      .asScala
+      .zip(hiveSchema.getAllStructFieldTypeInfos.asScala.map(_.getTypeName))
+      .map(_.productIterator.mkString(": "))
+      .mkString("\n")
     new MetaException(
       s"""The Delta table schema is not the same as the Hive schema. Please update your Hive
          |table's schema to match the Delta table schema.
-         |Delta table schema: $deltaSchema
-         |Hive schema: $hiveSchema""".stripMargin)
+         |
+         |Delta table schema:
+         |${deltaSchema.treeString}
+         |
+         |Hive schema:
+         |${hiveSchemaString}""".stripMargin)
   }
 
   // TODO Configure `spark` to pick up the right Hadoop configuration.

--- a/hive/src/test/scala/io/delta/hive/HiveConnectorSuite.scala
+++ b/hive/src/test/scala/io/delta/hive/HiveConnectorSuite.scala
@@ -547,6 +547,55 @@ class HiveConnectorSuite extends HiveTest with BeforeAndAfterEach {
       }
     }
   }
+
+  test("column names should be case insensitive") {
+    // Create a Delta table
+    withTable("deltaCaseInsensitiveTest") {
+      withTempDir { dir =>
+        val testData = (0 until 10).map(x => (x, s"foo${x % 2}"))
+
+        withSparkSession{ spark =>
+          import spark.implicits._
+          testData.toDS.toDF("FooBar", "BarFoo").write.format("delta")
+            .partitionBy("BarFoo").save(dir.getCanonicalPath)
+        }
+
+        runQuery(
+          s"""
+             |create external table deltaCaseInsensitiveTest(fooBar int, Barfoo string)
+             |stored by 'io.delta.hive.DeltaStorageHandler' location '${dir.getCanonicalPath}'
+         """.stripMargin
+        )
+
+        checkAnswer("select * from deltaCaseInsensitiveTest", testData)
+        for ((col1, col2) <-
+               Seq("fooBar" -> "barFoo", "foobar" -> "barfoo", "FOOBAR" -> "BARFOO")) {
+          checkAnswer(
+            s"select $col1, $col2 from deltaCaseInsensitiveTest",
+            testData)
+          checkAnswer(
+            s"select $col2, $col1 from deltaCaseInsensitiveTest",
+            testData.map(_.swap))
+          checkAnswer(
+            s"select $col1 from deltaCaseInsensitiveTest where $col2 = '2'",
+            testData.filter(_._2 == "2").map(x => OneItem(x._1)))
+          checkAnswer(
+            s"select $col2 from deltaCaseInsensitiveTest where $col1 = 2",
+            testData.filter(_._1 == 2).map(x => OneItem(x._2)))
+        }
+        for (col <- Seq("fooBar", "foobar", "FOOBAR")) {
+          checkAnswer(
+            s"select $col from deltaCaseInsensitiveTest",
+            testData.map(x => OneItem(x._1)))
+        }
+        for (col <- Seq("barFoo", "barfoo", "BARFOO")) {
+          checkAnswer(
+            s"select $col from deltaCaseInsensitiveTest",
+            testData.map(x => OneItem(x._2)))
+        }
+      }
+    }
+  }
 }
 
 case class TestStruct(f1: String, f2: Long)
@@ -569,3 +618,5 @@ case class TestClass(
   c14: Map[String, Long],
   c15: TestStruct
 )
+
+case class OneItem[T](t: T)

